traffic_ops/traffic_ops_golang/routes.go

```diff
@@ -381,6 +374,10 @@ func Routes(d ServerData) ([]Route, []RawRoute, http.Handler, error) {
 		{1.1, http.MethodDelete, `deliveryservices/{id}/?(\.json)?$`, api.DeleteHandler(deliveryservice.GetTypeV12Factory()), auth.PrivLevelOperations, Authenticated, nil},
 		{1.1, http.MethodGet, `deliveryservices/{id}/servers/eligible/?(\.json)?$`, deliveryservice.GetServersEligible, auth.PrivLevelReadOnly, Authenticated, nil},
 
+		{1.1, http.MethodGet, `deliveryservices/xmlId/{xmlid}/sslkeys$`, deliveryservice.GetSSLKeysByXMLID, auth.PrivLevelAdmin, Authenticated, nil},
+		{1.1, http.MethodGet, `deliveryservices/hostname/{hostname}/sslkeys$`, deliveryservice.GetSSLKeysByHostName, auth.PrivLevelAdmin, Authenticated, nil},
+		{1.1, http.MethodPost, `deliveryservices/sslkeys/add$`, deliveryservice.AddSSLKeys, auth.PrivLevelAdmin, Authenticated, nil},
+		{1.1, http.MethodGet, `deliveryservices/xmlId/{xmlid}/sslkeys/delete$`, deliveryservice.DeleteSSLKeys, auth.PrivLevelOperations, Authenticated, nil},
```

Comment:
This doesn't seem to work correctly. I hit this endpoint with `?version=6` (where that version existed), and it returned `Successfully deleted ssl key`, but the key wasn't deleted, it still existed and was still returned with `GET ?latest`. 

I then called the delete endpoint with `?version=latest`, and it deleted all keys on that delivery service. Which I did _not_ expect. I'm not sure if Perl does that, but I expected it to delete the latest key, definitely not all keys (presumably because it couldn't find that version?).

Reply:
So I believe what you're describing is expected behavior. Whenever certs are added to a DS, they get stored in riak under two different keys:
`ssl/<xml_id>-<version>` where `<version>` is an int and matches the version inside the record
`ssl/<xml_id>-latest`
So the first certs added to a DS get stored in Riak under these keys:
`ssl/myds-1`
`ssl/myds-latest`
The Riak keys are different, but the underlying records are the same.
When another set of certs is added to the DS via Traffic Portal, Traffic Portal increments the version from the existing record and passes the incremented version in the request. The new certs then get added to Riak under:
`ssl/myds-2`
`ssl/myds-latest` <- this _overwrites_ the existing "latest" record.
By default, when you don't specify a version for GET/DELETE, it will choose the `ssl/myds-latest` record.

So when you deleted version 6, you really deleted `ssl/myds-6`, which leaves `ssl/myds-latest` in Riak still. When you then deleted version "latest", you deleted `ssl/myds-latest`. I believe if you do a GET with `?version=5` it should return `ssl/myds-5` which should still exist as long as it was created.

Reply:
Let's say "designed", not "expected" :P. I'll double-check the `/delete` endpoint, but if our tools and UI behave that way, I guess it's fine.

>When you then deleted version "latest", you deleted ssl/myds-latest. I believe if you do a GET with ?version=5

Yes, that's what happened (and the previous version I tried, 3, actually didn't exist, making me think they'd all been deleted). But that leaves the SSL keys in an unusable state for our tools, because everything gets `latest`, and the API even returns that (=not found) when the version isn't specified.

It deeply unsettles me that I expected to `delete?version=x` and delete that version in the db, and have "latest" now be the prior, and likewise for `delete?version=latest` to delete the actual key, under both the riak key and the real key. I very much did *not* expect `/delete?version=latest` to delete a copied key, appear to have deleted all the keys, and break all of TC that uses it.

In both cases, I expected the version queries to get and act on the real key, not for there to be multiple keys actually stored, and not for `delete?latest` to delete the key stored there (breaking all our tools), and not the actual latest key, and for `?latest` to always return the real latest key in the db. If I delete `version=6`, I expect `?latest` to return version 5. If I delete `version=latest`, I expect version 6 to be deleted, and `?latest` to return 5.

Again, I won't hold up this PR if `/delete` works as designed. But I seriously think we should rethink our design. If anyone else misunderstands or shares my intuitive expectations, it could lead to some serious production problems for them.

Reply:
I agree with everything you've described here. I opened #2877 while I was working on this PR to track how we think we can make these APIs better in TO API 2.0. I will try to capture what we've discussed here in that issue so that we can make this API better in TO API 2.0.